dpd/p4/sidecar.p4

```diff
@@ -56,34 +138,60 @@ control Filter(
 	// Table of the IPv6 addresses assigned to ports on the switch.
 	table switch_ipv6_addr {
 		key = {
-			hdr.ipv6.dst_addr: exact;
-			meta.in_port: ternary;
+			hdr.ipv6.dst_addr : exact;
+			ig_intr_md.ingress_port : ternary;
 		}
 		actions = { claimv6; dropv6; }
 
 		const size = SWITCH_IPV6_ADDRS_SIZE;
 		counters = ipv6_ctr;
 	}
 
-
 	apply {
-		if (hdr.ipv4.isValid() || hdr.arp.isValid()) {
+		if (hdr.arp.isValid()) {
+			switch_ipv4_addr.apply();
+		} else if (hdr.ipv4.isValid()) {
+			// Check if this is an IPv4 multicast address
+			bit<16> mcast_scope_v4 = (bit<16>)hdr.ipv4.dst_addr[31:28];
+			if (mcast_scope_v4 == 16w0xe) {
+				bool is_valid = validate_ipv4_mcast_mac(hdr.ethernet.dst_mac, hdr.ipv4.dst_addr);
+				if (is_valid) {
+					meta.is_mcast = true;
+				} else {
+					drop_invalid_mcast_mac();
+					drop_mcast_ctr.count(ig_intr_md.ingress_port);
+					drop_reason_ctr.count(meta.drop_reason);
+				}
+			}
 			switch_ipv4_addr.apply();
 		} else if (hdr.ipv6.isValid()) {
-			multicast_scope = (bit<16>)hdr.ipv6.dst_addr[127:112];
-			if (multicast_scope == 16w0xff01) {
-				ig_dprsr_md.drop_ctl = 1;
-				meta.drop_reason = DROP_MULTICAST_TO_LOCAL_INTERFACE;
-			} else {
-				meta.multicast = (multicast_scope == 16w0xff02);
-				switch_ipv6_addr.apply();
+			bit<16> mcast_scope_v6 = (bit<16>)hdr.ipv6.dst_addr[127:112];
+			if (mcast_scope_v6 == 16w0xff01) {
+				// Interface-local multicast
+				drop_mcast_local_interface();
+				drop_mcast_ctr.count(ig_intr_md.ingress_port);
+				drop_reason_ctr.count(meta.drop_reason);
+			} else if (hdr.ipv6.dst_addr[127:120] == 8w0xff) {
+				bool is_valid = validate_ipv6_mcast_mac(hdr.ethernet.dst_mac, hdr.ipv6.dst_addr);
+				if (is_valid) {
+					if (mcast_scope_v6 == 16w0xff02) {
+						// IPv6 Link-local multicast
+						meta.is_link_local_mcast = true;
```

Comment:
Is the reason behind all these derived values in metadata to avoid issues with PHV allocation when using actual packet values as match keys?

Reply:
Yep. Going back to the ingress model, I could make these meta again, but if we add more meta fields, we'll run up against it, which I did.